### PR TITLE
feat: updatable validator commissions

### PR DESCRIPTION
**Pull request by sveitser**
- stake table contract: allow updates with rate limits
- migrate commissions from V1 stake table into storage for V2
- handle commission changes in the GCL
- test: stake table state gets invalid commission
- break up some lines to fix cargo fmt
- handle patch upgrades for stake table
- stake table upgrade: dedupe upgrade preparation
- test: unittest commission updates

### What's not done
- Enforcing of limits to commission changes in GCL. I left a comment, not quite clear to me if we want to do that since it doesn't seem relevant for consensus.
- Add a new protocol version. I think we may be able to do this but it's pending some discussion.

Tests:
```
forge test --match-contract StakeTableV2Comm
cargo test -p espresso-types
cargo test -p espresso-contract-deployer
```

### CODE REVIEW

File: contracts/rust/deployer/src/lib.rs [hunk 1]
```diff
@@ -1568,6 +1678,133 @@ mod tests {
...
+            &provider,
+            *stake_table_v2.address(),
+        )
+        .await
+        .unwrap_err();
+        assert!(err.to_string().contains("Expected StakeTable V1"));
+
+        Ok(())
+    }
+
```
Comment (sveitser) [inline contracts/rust/deployer/src/lib.rs:1769]: Should add a test that does a deployment with (non-empty) commission migration.

File: contracts/src/StakeTableV2.sol [hunk 2]
```diff
@@ -55,9 +55,47 @@ import { SafeTransferLib } from "solmate/utils/SafeTransferLib.sol";
...
+///
 /// @notice The StakeTableV2 contract ABI is a superset of the original ABI. Consumers of the
 /// contract can use the V2 ABI, even if they would like to maintain backwards compatibility.
 contract StakeTableV2 is StakeTable, PausableUpgradeable, AccessControlUpgradeable {
+    // === Types ===
+
+    /// @notice Struct for tracking validator commission and last increase time
+    struct CommissionTracking {
+        uint16 commission;
+        uint256 lastIncreaseTime;
```
Comment (alysiahuggins) [inline contracts/src/StakeTableV2.sol:68]: a validator may also decrease commission so maybe rename to `lastModified` or `lastChanged`

Comment (sveitser) [reply contracts/src/StakeTableV2.sol:68]: The decreases are not rate limited, and therefore the timestamp isn't tracked and and the amount of the decrease isn't checked.

File: contracts/rust/deployer/src/lib.rs (hunk 1 above)
Comment (sveitser) [reply contracts/rust/deployer/src/lib.rs:1769]: for test added here: 1bc21809e13a3802dad082e07f476336de99382c

File: contracts/src/StakeTableV2.sol (hunk 2 above)
Comment (alysiahuggins) [reply contracts/src/StakeTableV2.sol:68]: ok gotcha

File: contracts/src/StakeTableV2.sol [hunk 3]
```diff
@@ -55,9 +55,47 @@ import { SafeTransferLib } from "solmate/utils/SafeTransferLib.sol";
...
+        address validator;
+        uint16 commission;
+    }
+
+    // === Storage ===
+
+    bytes32 public constant PAUSER_ROLE = keccak256("PAUSER_ROLE");
+
+    /// @notice Minimum time interval between commission updates (in seconds)
+    uint256 public minCommissionUpdateInterval;
```
Comment (alysiahuggins) [inline contracts/src/StakeTableV2.sol]: should this be `minCommissionUpdateInterval` since it only applies to commission increases as only increases are rate limited as mentioned [here](https://github.com/EspressoSystems/espresso-network/pull/3591/files#r2329362640). 

also maybe this can be uint32?

File: contracts/src/StakeTableV2.sol [hunk 4]
```diff
@@ -290,6 +373,93 @@ contract StakeTableV2 is StakeTable, PausableUpgradeable, AccessControlUpgradeab
...
+    /// These limits protect stakers from sudden large commission increases,
+    /// particularly by exiting validators.
+    function updateCommission(uint16 newCommission) external virtual whenNotPaused {
+        address validator = msg.sender;
+        ensureValidatorActive(validator);
+        require(newCommission <= 10000, InvalidCommission());
+
+        CommissionTracking storage tracking = commissionTracking[validator];
+        uint16 currentCommission = tracking.commission;
+        uint256 lastIncreaseTime = tracking.lastIncreaseTime;
```
Comment (alysiahuggins) [inline contracts/src/StakeTableV2.sol]: you can put this line within the for loop as this is the only place where `lastIncreaseTime` is used

File: contracts/src/StakeTableV2.sol [hunk 5]
```diff
@@ -290,6 +373,93 @@ contract StakeTableV2 is StakeTable, PausableUpgradeable, AccessControlUpgradeab
...
+        if (newCommission > currentCommission) {
+            // Allow immediate first increase or after interval has passed
+            require(
+                lastIncreaseTime == 0
+                    || block.timestamp >= lastIncreaseTime + minCommissionUpdateInterval,
+                CommissionUpdateTooSoon()
+            );
+
+            uint16 increase = newCommission - currentCommission;
+            require(increase <= maxCommissionIncrease, CommissionIncreaseExceedsMax());
```
Comment (alysiahuggins) [inline contracts/src/StakeTableV2.sol]: i'm thinking we can remove the line above and change this to
```suggestion
           if (newCommission > currentCommission + maxCommissionIncrease) revert CommissionIncreaseExceedsMax();
```

File: contracts/src/StakeTableV2.sol [hunk 6]
```diff
@@ -290,6 +373,93 @@ contract StakeTableV2 is StakeTable, PausableUpgradeable, AccessControlUpgradeab
...
+        // to commission increases.
+        //
+        // 1. Limits only enforced for commission *increase*.
+        // 2. Time of last change only tracked for commission *increase*.
+        if (newCommission > currentCommission) {
+            // Allow immediate first increase or after interval has passed
+            require(
+                lastIncreaseTime == 0
+                    || block.timestamp >= lastIncreaseTime + minCommissionUpdateInterval,
+                CommissionUpdateTooSoon()
```
Comment (alysiahuggins) [inline contracts/src/StakeTableV2.sol]: same nit as above about the use of require
```suggestion
            if(
                lastIncreaseTime != 0
                    && block.timestamp < lastIncreaseTime + minCommissionUpdateInterval)
                revert CommissionUpdateTooSoon()
```

File: contracts/src/StakeTableV2.sol [hunk 7]
```diff
@@ -84,6 +122,22 @@ contract StakeTableV2 is StakeTable, PausableUpgradeable, AccessControlUpgradeab
     /// @notice The exit escrow period is updated
     event ExitEscrowPeriodUpdated(uint64 newExitEscrowPeriod);
 
+    /// @notice A validator updates their commission rate
+    /// @param validator The address of the validator
+    /// @param timestamp The timestamp of the update
+    /// @param newCommission The new commission rate
+    ///
+    /// @dev the timestamp is emitted to simplify processing in the GCL
+    event CommissionUpdated(address indexed validator, uint256 timestamp, uint16 newCommission);
```
Comment (alysiahuggins) [inline contracts/src/StakeTableV2.sol]: could be useful to log the previous commission too

File: contracts/src/StakeTableV2.sol [hunk 8]
```diff
@@ -290,6 +373,93 @@ contract StakeTableV2 is StakeTable, PausableUpgradeable, AccessControlUpgradeab
...
+    function updateCommission(uint16 newCommission) external virtual whenNotPaused {
+        address validator = msg.sender;
+        ensureValidatorActive(validator);
+        require(newCommission <= 10000, InvalidCommission());
+
+        CommissionTracking storage tracking = commissionTracking[validator];
+        uint16 currentCommission = tracking.commission;
+        uint256 lastIncreaseTime = tracking.lastIncreaseTime;
+
+        require(newCommission != currentCommission, CommissionUnchanged());
```
Comment (alysiahuggins) [inline contracts/src/StakeTableV2.sol:394]: random nit: in our codebase, we normally write it like 
```suggestion
        if(newCommission == currentCommission) revert CommissionUnchanged();
```

File: contracts/src/StakeTableV2.sol [hunk 9]
```diff
@@ -290,6 +373,93 @@ contract StakeTableV2 is StakeTable, PausableUpgradeable, AccessControlUpgradeab
...
+    }
+
+    /// @notice Initialize validator commissions during V2 migration
+    /// @dev This function is used to retroactively initialize commission storage for validators
+    /// that were registered before the V2 upgrade. On decaf, this will be called with current
+    /// commission values read from L1 events. On mainnet, this will be called with an empty array
+    /// since there are no pre-existing validators.
+    /// @param initialCommissions Array of InitialCommission structs containing validator addresses
+    /// and their commissions
+    function _initializeCommissions(InitialCommission[] calldata initialCommissions) private {
```
Comment (alysiahuggins) [inline contracts/src/StakeTableV2.sol:446]: before we start the for loop, we should do a check to see if the length of initialCommissions is equal to the number of validators in the contract, and revert otherwise

File: contracts/test/StakeTableV2Commission.t.sol [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+        vm.stopPrank();
+    }
+
+    function test_DefaultValues() public view {
+        // This is the only test that checks default values - if defaults change, only this test
+        // should fail
+        assertEq(proxy.minCommissionUpdateInterval(), 7 days);
+        assertEq(proxy.maxCommissionIncrease(), 500);
+    }
+}
```
Comment (alysiahuggins) [inline contracts/test/StakeTableV2Commission.t.sol:268]: we should add a test for handling an upgrade from v1 to v2 where the initial commission array is not empty

File: contracts/src/StakeTableV2.sol (hunk 3 above)
Comment (sveitser) [reply contracts/src/StakeTableV2.sol]: I'll change the name. Yeah it could be a smaller type but I think just using the same type as block.timestamp is okay.

File: contracts/src/StakeTableV2.sol (hunk 6 above)
Comment (sveitser) [reply contracts/src/StakeTableV2.sol]: Why avoid require?

File: contracts/src/StakeTableV2.sol (hunk 8 above)
Comment (sveitser) [reply contracts/src/StakeTableV2.sol:394]: I think this is for historical reasons because require did not support custom errors when we started. It supports that now.

File: contracts/test/StakeTableV2Commission.t.sol (hunk 10 above)
Comment (sveitser) [reply contracts/test/StakeTableV2Commission.t.sol:268]: There are rust tests for this (which also checks that we are correctly fetching the commissions), but sure can also add a solidity test.

File: contracts/src/StakeTableV2.sol (hunk 6 above)
Comment (alysiahuggins) [reply contracts/src/StakeTableV2.sol]: mostly because we use the `if (!cond) revert Error();` format in the rest of the codebase

File: contracts/src/StakeTableV2.sol (hunk 9 above)
Comment (sveitser) [reply contracts/src/StakeTableV2.sol:446]: Uhm... we can't get the number of validators from the validators mapping and don't store it otherwise.

Comment (sveitser) [reply contracts/src/StakeTableV2.sol:446]: I think this is okay though, the migration will only be applied on decaf. There are checks that the validators are registered so we can't accidentally pass anything on the fresh mainnet deployment.

Comment (sveitser) [reply contracts/src/StakeTableV2.sol:446]: There is a problem though if the validator has zero commission, then after initialization it's not possible to tell if the tracking is initialized or not.

File: contracts/src/StakeTableV2.sol (hunk 5 above)
Comment (sveitser) [reply contracts/src/StakeTableV2.sol]: partially done in https://github.com/EspressoSystems/espresso-network/pull/3591/commits/6e6f3b2ea1ff644b9cba7836dd863f8fb7910562

File: contracts/src/StakeTableV2.sol (hunk 6 above)
Comment (sveitser) [reply contracts/src/StakeTableV2.sol]: I think using require is fine. It reads better. I think the use of if ... revert just became common when custom errors weren't supported with require.

File: contracts/src/StakeTableV2.sol (hunk 7 above)
Comment (sveitser) [reply contracts/src/StakeTableV2.sol]: https://github.com/EspressoSystems/espresso-network/pull/3591/commits/6e6f3b2ea1ff644b9cba7836dd863f8fb7910562

File: contracts/src/StakeTableV2.sol (hunk 9 above)
Comment (sveitser) [reply contracts/src/StakeTableV2.sol:446]: Added more tests in https://github.com/EspressoSystems/espresso-network/pull/3591/commits/6e6f3b2ea1ff644b9cba7836dd863f8fb7910562

Comment (sveitser) [reply contracts/src/StakeTableV2.sol:446]: So basically if a validator has zero commission it can be duplicated. I think this is fine as long as our rust code fetches and prepares the commissions correctly which I think it does.

File: contracts/src/StakeTableV2.sol (hunk 4 above)
Comment (sveitser) [reply contracts/src/StakeTableV2.sol]: https://github.com/EspressoSystems/espresso-network/commit/6e6f3b2ea1ff644b9cba7836dd863f8fb7910562

File: contracts/src/StakeTableV2.sol (hunk 3 above)
Comment (sveitser) [reply contracts/src/StakeTableV2.sol]: renamed in https://github.com/EspressoSystems/espresso-network/commit/6e6f3b2ea1ff644b9cba7836dd863f8fb7910562

**Comment by alysiahuggins** [comment]
the contract side of things lgtm

File: contracts/rust/deployer/src/lib.rs [hunk 11]
```diff
@@ -1568,6 +1686,183 @@ mod tests {
...
+    /// env RPC_URL=... cargo test -p espresso-contract-deployer -- --ignored test_fetch_commissions_sepolia
+    #[ignore]
+    #[test_log::test(tokio::test)]
+    async fn test_fetch_commissions_sepolia() -> Result<()> {
+        let rpc_url = std::env::var("RPC_URL")
+            .expect("RPC_URL environment variable not set")
+            .parse()?;
+        let provider = ProviderBuilder::new().on_http(rpc_url);
+
+        let stake_table_address: Address = "0x40304FbE94D5E7D1492Dd90c53a2D63E8506a037".parse()?;
```
Comment (imabdulbasit) [inline contracts/rust/deployer/src/lib.rs:1785]: please add a comment that this is decaf contract address

File: sequencer/src/api.rs [hunk 12]
```diff
@@ -5684,6 +5706,130 @@ mod test {
...
+            .unwrap();
+        let target_epoch = current_epoch.u64() + 3;
+        let mut events = network.peers[0].event_stream().await;
+        wait_for_epochs(&mut events, EPOCH_HEIGHT, target_epoch).await;
+
+        // the last epoch with the old commissions
+        let client: Client<ServerError, SequencerApiVersion> =
+            Client::new(format!("http://localhost:{api_port}").parse().unwrap());
+        let validators = client
+            .get::<ValidatorMap>(&format!("node/validators/{}", target_epoch - 1))
```
Comment (imabdulbasit) [inline sequencer/src/api.rs:5796]: Shouldn’t this quer target_epoch - 2? Since a contract update only takes two epochs to take effect, there’s a possibility of overlap

File: sequencer/src/api.rs [hunk 13]
```diff
@@ -5684,6 +5706,130 @@ mod test {
...
+            .await
+            .expect("validators");
+        assert!(!validators.is_empty());
+        for (val, _, new_comm) in commissions.clone() {
+            assert_eq!(validators.get(&val).unwrap().commission, new_comm.to_evm());
+        }
+
+        // check validator 0 received zero rewards in the last epoch where its
+        // commission was zero
+        for block in EPOCH_HEIGHT * (target_epoch - 1) + 1..EPOCH_HEIGHT * target_epoch {
```
Comment (imabdulbasit) [inline sequencer/src/api.rs]: can we also just assert that other two validators received some rewards?

File: staking-cli/src/deploy.rs [hunk 14]
```diff
@@ -65,11 +65,25 @@ impl TestSystem {
...
+        let mut attempts = 0;
+        let (port, provider) = loop {
+            let port = portpicker::pick_unused_port().unwrap();
+            match ProviderBuilder::new().on_anvil_with_wallet_and_config(|anvil| {
+                anvil.port(port).arg("--accounts").arg("20")
+            }) {
+                Ok(provider) => break (port, provider),
+                Err(e) => {
+                    attempts += 1;
+                    if attempts >= 5 {
```
Comment (imabdulbasit) [inline staking-cli/src/deploy.rs:81]: should there be some delay before retrying like 1 s sleep?

File: staking-cli/src/registration.rs [hunk 15]
```diff
@@ -210,6 +243,44 @@ mod test {
         Ok(())
     }
 
+    #[tokio::test]
+    async fn test_update_commission() -> Result<()> {
```
Comment (imabdulbasit) [inline staking-cli/src/registration.rs:247]: I think we should also test that a commission update fails if it doesn’t respect the update interval

File: types/src/v0/impls/stake_table.rs [hunk 16]
```diff
@@ -503,6 +511,27 @@ impl StakeTableState {
                 validator.stake_table_key = stake_table_key;
                 validator.state_ver_key = state_ver_key;
             },
+
+            StakeTableEvent::CommissionUpdate(CommissionUpdated {
```
Comment (imabdulbasit) [inline types/src/v0/impls/stake_table.rs:516]: should add comments that we don't expect commission update events for protocol version V3 etc

File: types/src/v0/impls/stake_table.rs [hunk 17]
```diff
@@ -1034,7 +1066,7 @@ impl Fetcher {
             }
         });
         // fetch consensus keys updated v2 events
-        let keys_update_events_v2 = stream::iter(chunks).then(|(from, to)| {
+        let keys_update_events_v2 = stream::iter(chunks.clone()).then(|(from, to)| {
```
Comment (imabdulbasit) [inline types/src/v0/impls/stake_table.rs:1072]: why does it require `clone` now?

File: types/src/v0/impls/stake_table.rs [hunk 18]
```diff
@@ -1500,9 +1564,10 @@ impl EpochCommittees {
     /// We used a fixed block reward for version v3
     /// Version v4 uses the dynamic block reward based
     /// Assumes the stake table contract proxy address does not change
-    /// In the future, if we want to support updates to the stake table contract address via chain config,
-    /// or allow the contract to handle additional block reward calculation parameters (e.g., inflation, block time),
-    /// the `fetch_block_reward` logic can be updated to support per-epoch rewards.
+    /// In the future, if we want to support updates to the stake table contract address
+    /// via chain config, or allow the contract to handle additional block reward calculation
+    /// parameters (e.g., inflation, block time), the `fetch_block_reward` logic can be
```
Comment (imabdulbasit) [inline types/src/v0/impls/stake_table.rs]: this comment is outdated now.

File: staking-cli/src/deploy.rs (hunk 14 above)
Comment (sveitser) [reply staking-cli/src/deploy.rs:81]: I don't really know what the underlying issue is so I'm not sure waiting has any positive effect. I haven't seen things fail after making this change.

File: staking-cli/src/registration.rs (hunk 15 above)
Comment (sveitser) [reply staking-cli/src/registration.rs:247]: This is tested in the solidity tests but I'll add it here as well.

File: types/src/v0/impls/stake_table.rs (hunk 17 above)
Comment (sveitser) [reply types/src/v0/impls/stake_table.rs:1072]: because we use it again later, but you're right the last use doesn't require clone

File: sequencer/src/api.rs (hunk 13 above)
Comment (sveitser) [reply sequencer/src/api.rs]: ah, forgot this

Comment (sveitser) [reply sequencer/src/api.rs]: aec6dae25fdf2c812d657c05b598a3ff5ba2addb

File: staking-cli/src/parse.rs [hunk 19]
```diff
@@ -126,7 +130,11 @@ mod test {
             ("2", 200),
             ("1.000000", 100),
             ("1.2", 120),
+            ("1.2", 120),
```
Comment (imabdulbasit) [inline staking-cli/src/parse.rs]: looks the same as above row ?

Comment (sveitser) [reply staking-cli/src/parse.rs]: https://github.com/EspressoSystems/espresso-network/pull/3591/commits/d69c930183497029a2bec7dbc13a4ae2fc9092db

File: sequencer/src/api.rs (hunk 12 above)
Comment (sveitser) [reply sequencer/src/api.rs:5796]: @imabdulbasit you mean in case the commission changes happen between the epoch root block and the final block in an epoch?

Comment (imabdulbasit) [reply sequencer/src/api.rs:5796]: Let’s say we are in epoch 4. We updated the commission in epoch 4. The target epoch = 4 + 3 = 7. Epoch 6 will still have the old commission because it takes 2 full epochs?

Comment (imabdulbasit) [reply sequencer/src/api.rs:5796]: so I think it should be fine actually
